Create serialization.txt (#1893)
mongodb/docs-realm: source/sdk/kotlin/realm-database/serialization.txt

.. _kotlin-serialization:

==========================
Serialization - Kotlin SDK
==========================

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Serialization methods used by libraries that depend on reflection, such as
`GSON <https://github.com/google/gson>`__ do not work with the SDK
by default.

> 

Reviewer: issue (non-blocking): This article may be more valuable to the reader by placing the code to serialize earlier with a brief explanation.

```suggestion
You must exclude certain hidden fields from serialization when using external libraries like GSON.

<code>

The SDK compiler plugin injects (a hidden field|hidden fields)[how many?]...
```

> This happens because the SDK compiler plugin injects a hidden field

Reviewer 2: Nit: the use of "happens" here reads a bit odd to me. I would expect to see that after reading about a specific type of failure or error message, not a more general "does not work." Maybe:

> This is because...

Or just:

> The SDK compiler plugin...

Author: ```suggestion
This is because the SDK compiler plugin injects a hidden field
```

.. _kotlin-serialization:

==========================

> Serialization - Kotlin SDK

Reviewer 3: Having a section describing the internals in a general way like this sounds like a good idea 👍

==========================

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

> The serialization methods used by libraries such as
> `GSON <https://github.com/google/gson>`__ do not work with the SDK
> by default.

Reviewer 3: ```suggestion
Serialization methods used by libraries that depend on reflection, such as
`GSON <https://github.com/google/gson>`__ do not work with the SDK
by default.
```

Added explanation making it clear what the problem is.

> This happens because the SDK compiler plugin injects hidden fields
> into object models, prefixed with ``io_realm_kotlin_``. The SDK uses
> these hidden fields to manage internal object state. Any library that
> relies on fields instead of getters and setters needs to ignore these
> hidden fields.

Reviewer 3: ```suggestion
This happens because the SDK compiler plugin injects a hidden field
into object models, prefixed with ``io_realm_kotlin_``. The SDK uses
this hidden field to manage internal object state. Any library that
relies on fields instead of getters and setters needs to ignore this
hidden field.
```

We only have 1 field, so rephrased to match.

To use the SDK with external libraries such as GSON, exclude the hidden
fields from serialization using a prefix match:

.. code-block:: kotlin
   :copyable: false

   var gson: Gson = GsonBuilder()

>        .registerTypeAdapter(Date::class.java, CustomDateTypeAdapter())

Reviewer 3: ```suggestion
```

This was only used by the project exposing this, so isn't relevant to this particular problem.
